### DISCUSSION TITLE
What is the strategy to verify the change before commiting?

### DISCUSSION THREAD

**Discussion by XtremeAI** [fluxcd/flux2#1617] [Q&A]
With Flux and its pruning mechanics it is easy to forget / make a mistake and get some resources removed from a cluster or created them in the wrong way.

Once I commit the change I see the problems straight away in Slack notifications but I would rather see these messages upfront. 

Is there a strategy to verify the changes Flux will apply beforehand, like i.e. Terraform does?

**Comment by hiddeco** [comment]
To verify changes before they are committed to your main repository, you can use the validation script here: https://github.com/fluxcd/flux2-kustomize-helm-example/blob/main/scripts/validate.sh

**Reply by XtremeAI** [reply]
Sorry, the script I believe only validates that the source can be compiled by kustomize and produces valid kubernetes manifests that can be applied to the cluster. 

I already do that but that is now what I am refferring to. I want to have some kind of a diff document that would show exactly what Flux will be **creating**, **deleting** or **modifying**. 

This it what Terraform does. I see straight away '-' will be removed, '+' will be added and '~' will be modified:  
```
  # module.main.module.flux.kubectl_manifest.flux_ks_tenant_manifest will be updated in-place
  ~ resource "kubectl_manifest" "flux_ks_tenant_manifest" {
        id                      = "/apis/kustomize.toolkit.fluxcd.io/v1beta1/namespaces/flux-system/kustomizations/tenant"
        name                    = "tenant"
      ~ yaml_body               = (sensitive value)
      ~ yaml_body_parsed        = <<-EOT
            apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
            kind: Kustomization
            metadata:
              name: tenant
              namespace: flux-system
            spec:
              decryption:
                provider: sops
              dependsOn:
              - name: platform
              interval: 1m0s
          -   path: ./env/cn
          +   path: ./env/aws-cn-dev
```

**Reply by phillebaba** [reply]
I think I understand what you are asking for, but I feel like this is was git already does when you create a PR and diff the changes. The changes that Flux will do is that changes that you commit to your repository. Things obviously get a lot more complicated when working with Helm.

Or are you looking for a command similar to `terraform plan` which would show the changes that will be applied if something is committed?

**Reply by rupelu** [reply]
This would make a really great feature and it would leverage PRs to talk about the changes in more detail. We have also the case that with pruning in place, commenting out a line could lead to several services getting decomissioned. This is great, but only if it is really intended.
It would be great to be able to do some kind of helm upgrade --install --dry-run or something similar to terraform plan just with flux. 
We use the same repository for multiple clusters so it would also be great if the flux-cli could have such a feature in a way that it can be pointed to a cluster and things get templated. However, it is quite hard to do this for helm, especially when using things like lookups and so on.
Our current solution to find out the changes is to have a full cluster in between that gets the changeset first.

**Reply by phillebaba** [reply]
This is kind of the good and the bad with using Helm. It does a lot for you which can mean adding a bunch of resources or removing a bunch of resources. I do not know how easy this would be to implement, partially because it would require access to your Helm repositories but also because it would go against the core point of Flux by exposing the k8s api to your CD solution.

I guess you could do this with some `yq` hacking by extracting the values, repository, and chart name from your helm release resource and then running helm template. With that you could diff it against what is currently in master. I am just not sure if this is something that should be implemented in the flux cli.

**Reply by anwarchk** [reply]
I think a client side feature to find the diffs on current state from Git and thats stored in the server would be a great features. Not to compare tools, but I have found https://carvel.dev/kapp/ to provide that ability. I know Flux is a pull based tool, but this would be great to include as part of the CLI

**Reply by anwarchk** [reply]
From Carvel's doc `kapp calculates changes between your configuration and live cluster state (a set of create, update, and delete operations) and applies changes you approve. It reports how resources are converging to their desired state providing insight into progress of the deployment.`

**Reply by XtremeAI** [reply]
> I think I understand what you are asking for, but I feel like this is was git already does when you create a PR and diff the changes. The changes that Flux will do is that changes that you commit to your repository. Things obviously get a lot more complicated when working with Helm.
> 
> Or are you looking for a command similar to `terraform plan` which would show the changes that will be applied if something is committed?

Sorry for a delayed response guys. 

I'll just top up on what @rupelu wrote. 

@phillebaba Yes, exactly, something like `terraform plan` would do the job way better than reviewing a diff on a PR. Especially in our case as we have a single Flux repo to manage a dozen of complex environments with prod and non-prod on it. I've tried to go the PR way and in our setup PRs turn to be very inefficient way of work.

I mentioned that already and will repeat it again. Flux already provides the info I need as alerts (post implementation). So if you could teach it to do a 'dry-run' and provide such alerts (on what is to be created / deleted (level of HelmRelease is enough)) before applying the changes (pre-commit hook? / flux cli feature?) - this would be a good solution! 

Having an exact cluster copy that gets the test change first - is an answer but a very costly one...

**Comment by mr-onion-2** [comment]
Sorry to drag up an old discussion but I wondered if there had been any recent developments to this or best practices?

_flux diff kustomization_ is perfect and I know it goes against the GitOps principles accessing the cluster from the CI, but does anyone run this this in their workflows prior to merging to master? Or is it very much frowned upon?  :)

**Reply by mloskot** [reply]
@mr-onion-2  Good question. I've just discovered the power of `flux diff kustomization` myself and I'm wondering the same.
I've found this thread oldish https://github.com/fluxcd/flux2/discussions/2783 which suggests there are other Flux users who attempt to run the command in their pipelines.

> `flux diff kustomization` is perfect and I know it goes against the GitOps principles accessing the cluster from the CI

As a GitOps newbie, this is a bit unclear to me. I'd expect read-only access to be excluded from the things that are against GitOps as falling into category of harvesting a cluster for status and resources details for reporting.

**Comment by jonnysw** [comment]
I'd be very keen to hear how other people are handling this too. We had planned to use `flux diff kustomization` in our pipeline for validating/verifying changes to our flux repository but it doesn't appear to be suitable as it only shows changes to the top level Kustomizations. However, these rarely ever change. The changes we are more concerned with are in the components that these Kustomizations apply.

For example, we have three cluster directories with top level Kustomizations in them for dev, uat and production, and each pull in a number of shared components, most of which are HelmReleases. If we make a change to one of these shared components, we want to know what affect that will have on each of the clusters via diff in the PR build, but I can't see any way to achieve this currently.

**Comment by varac** [comment]
I'm using https://github.com/tarioch/flux-check-hook and was happy with it, although it's last commit is 11 month old already.
